source/Calamari.Common/CalamariFlavourProgramAsync.cs

```diff
@@ -56,15 +56,14 @@ protected virtual void ConfigureContainer(ContainerBuilder builder, CommonOption
             builder.RegisterType<AssemblyEmbeddedResources>().As<ICalamariEmbeddedResources>();
             builder.RegisterType<ConfigurationVariablesReplacer>().As<IConfigurationVariablesReplacer>();
             builder.RegisterType<TransformFileLocator>().As<ITransformFileLocator>();
-            builder.RegisterType<JsonFormatVariableReplacer>().As<IFileFormatVariableReplacer>();
-            builder.RegisterType<YamlFormatVariableReplacer>().As<IFileFormatVariableReplacer>();
-            builder.RegisterType<StructuredConfigVariablesService>().As<IStructuredConfigVariablesService>();
             builder.Register(context => ConfigurationTransformer.FromVariables(context.Resolve<IVariables>(), context.Resolve<ILog>())).As<IConfigurationTransformer>();
             builder.RegisterType<DeploymentJournalWriter>().As<IDeploymentJournalWriter>().SingleInstance();
             builder.RegisterType<CodeGenFunctionsRegistry>().SingleInstance();
 
             var assemblies = GetAllAssembliesToRegister().ToArray();
 
+            builder.RegisterAssemblyTypes(assemblies).AssignableTo<IFileFormatVariableReplacer>().As<IFileFormatVariableReplacer>();
```

Comment:
I'm not sure why this is here. I think that the `StructuredConfigVariablesModule` ([below](https://github.com/OctopusDeploy/Calamari/pull/778/files#diff-04f72119f47dbc7a899939a63d263c455c3aa8cd09de9dd04dca3fa713f6e400R92)) should be handling this, in a way that respects the ordering of services.

Reply:
I _suspect_ this caters for the fact that we might want register file format variable replacers from an assembly that implements `CalamariFlavourProgramAsync` and overrides `GetProgramAssembliesToRegister`.

Caveat: I don't think we do this, nor do I think we really want to either. 

Also, I think the code here probably leads to double registrations, which I don't like. If we want to correct it for now and leave this capability in tact, I'd change this to

`builder.RegisterAssemblyTypes(GetProgramAssembliesToRegister())...`